src/shared/injector.js

```diff
@@ -134,4 +138,25 @@ export class Injector {
     this._providers.set(name, provider);
     return this;
   }
+
+  /**
+   * Run a function which uses one or more dependencies provided by the
+   * container.
+   *
+   * @param {Function} -
+   *   A callback to run, with dependencies annotated in the same way as
+   *   functions or classes passed to `register`.
+   * @return {any} - Returns the result of running the function.
+   */
+  run(callback) {
+    const tempName = 'Injector.run';
```

Comment:
Is it ever possible to have a state in which multiple `run`s are in flight, i.e. a namespace collision on `tempName`? e.g. if the function being run were async and another function got `run` in the meantime?

Reply:
If `callback` is async then `run` will still return synchronously but will return a Promise. However you have still found an issue. If `callback` happens to invoke `run` on the same container, that would be a problem as this function is not re-entrant. We could fix this by using a unique key on every call. I think we can live with this limitation for the moment though.